src/engine.rs

```diff
@@ -77,8 +78,12 @@ where
         let start = Instant::now();
         let mut builder = FilePipeLogBuilder::new(cfg.clone(), file_builder, listeners.clone());
         builder.scan()?;
-        let (append, rewrite) = builder.recover::<MemTableRecoverContext>()?;
+        let (_append, _rewrite) =
+            builder.recover::<MemTableRecoverContext>(&ReplayMachineBuilder::default())?;
```

Comment:
First, only accept a builder, no need to pass in extra generic argument like `MemTableRecoverContext`.
```
pub trait ReplayMachineFactory {
  type Machine: ReplayMachine;
  fn new_machine(&self) -> Self::Machine;
}
// for FilePipeLogBuilder
fn recover<F: ReplayMachineFactory>(factory: F) {
  // details //
}
```

Second, add a default factory implementation for those can be constructed via default:
```
struct DefaultMachineFactory<T: ReplayMachine + Default>;
impl<T: ReplayMachine + Default> ReplayMachineFactory for DefaultMachineFactory {
  type Machine = T;
  fn new_machine(&self) -> T { T::default() }
}
// in engine.rs
let (append, rewrite) = builder.recover(DefaultMachineFactory<MemTableRecoverContext>)?;
```